File: common/src/main/java/me/pandamods/pandalib/utils/animation/Easing.java
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU Lesser General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package me.pandamods.pandalib.utils.animation;

public interface Easing {
	double ease(double t);
	
	Easing easeInSine = t -> 1 - Math.cos((t * Math.PI) / 2);
	Easing easeOutSine = t -> Math.sin((t * Math.PI) / 2);
	Easing easeInOutSine = t -> -(Math.cos(Math.PI * t) - 1) / 2;
	
	Easing easeInQuad = t -> t * t;
	Easing easeOutQuad = t -> 1 - (1 - t) * (1 - t);
	Easing easeInOutQuad = t -> t < 0.5 ? 2 * t * t : 1 - (-2 * t + 2) * (-2 * t + 2) / 2;
	
	Easing easeInCubic = t -> t * t * t;
	Easing easeOutCubic = t -> 1 - Math.pow(1 - t, 3);
	Easing easeInOutCubic = t -> t < 0.5 ? 4 * t * t * t : 1 - Math.pow(-2 * t + 2, 3) / 2;
	
	Easing easeInQuart = t -> t * t * t * t;
	Easing easeOutQuart = t -> 1 - Math.pow(1 - t, 4);
	Easing easeInOutQuart = t -> t < 0.5 ? 8 * t * t * t * t : 1 - Math.pow(-2 * t + 2, 4) / 2;
	
	Easing easeInQuint = t -> t * t * t * t * t;
	Easing easeOutQuint = t -> 1 - Math.pow(1 - t, 5);
	Easing easeInOutQuint = t -> t < 0.5 ? 16 * t * t * t * t * t : 1 - Math.pow(-2 * t + 2, 5) / 2;
	
	Easing easeInExpo = t -> t == 0 ? 0 : Math.pow(2, 10 * (t - 1));
	Easing easeOutExpo = t -> t == 1 ? 1 : 1 - Math.pow(2, -10 * t);
	Easing easeInOutExpo = t -> t == 0 ? 0 : t == 1 ? 1 : t < 0.5 ? Math.pow(2, 20 * t - 10) / 2 : (2 - Math.pow(2, -20 * t + 10)) / 2;
	
	Easing easeInCirc = t -> 1 - Math.sqrt(1 - t * t);
	Easing easeOutCirc = t -> Math.sqrt(1 - Math.pow(t - 1, 2));
	Easing easeInOutCirc = t -> t < 0.5 ? (1 - Math.sqrt(1 - 4 * t * t)) / 2 : (Math.sqrt(1 - -4 * t * t) + 1) / 2;
	
	Easing easeInBack = t -> {
		double c1 = 1.70158;
		double c3 = c1 + 1;
		
		return c3 * t * t * t - c1 * t * t;
	};
	Easing easeOutBack = t -> {
		double c1 = 1.70158;
		double c3 = c1 + 1;
		
		return 1 + c3 * (t - 1) * (t - 1) * (t - 1) - c1 * (t - 1) * (t - 1);
	};
	Easing easeInOutBack = t -> {
		double c1 = 1.70158;
		double c2 = c1 * 1.525;

		return t < 0.5
				? (Math.pow(2 * t, 2) * ((c2 + 1) * 2 * t - c2)) / 2
				: (Math.pow(2 * t - 2, 2) * ((c2 + 1) * (t * 2 - 2) + c2) + 2) / 2;
	};
	
	Easing easeInElastic = t -> {
		double c4 = (2 * Math.PI) / 3;

		return t == 0
				? 0
				: t == 1
				? 1
				: -Math.pow(2, 10 * t - 10) * Math.sin((t * 10 - 10.75) * c4);
	};
	Easing easeOutElastic = t -> {
		double c4 = (2 * Math.PI) / 3;
		
		return t == 0
				? 0
				: t == 1
				? 1
				: Math.pow(2, -10 * t) * Math.sin((t * 10 - 0.75) * c4) + 1;
	};
	Easing easeInOutElastic = t -> {
		double c5 = (2 * Math.PI) / 4.5;

		return t == 0
				? 0
				: t == 1
				? 1
				: t < 0.5
				? -(Math.pow(2, 20 * t - 10) * Math.sin((20 * t - 11.125) * c5)) / 2
				: (Math.pow(2, -20 * t + 10) * Math.sin((20 * t - 11.125) * c5)) / 2 + 1;
	};

	Easing easeOutBounce = t -> {
		double n1 = 7.5625;
		double d1 = 2.75;

		if (t < 1 / d1) {
			return n1 * t * t;
		} else if (t < 2 / d1) {
			return n1 * (t -= 1.5 / d1) * t + 0.75;
		} else if (t < 2.5 / d1) {
			return n1 * (t -= 2.25 / d1) * t + 0.9375;
		} else {
			return n1 * (t -= 2.625 / d1) * t + 0.984375;
		}
	};
	Easing easeInBounce = t -> 1 - easeOutBounce.ease(1 - t);
	Easing easeInOutBounce = t -> t < 0.5 ? (1 - easeOutBounce.ease(1 - 2 * t)) / 2 : (1 + easeOutBounce.ease(2 * t - 1)) / 2;
}

File: src/main/kotlin/dev/pandasystems/pandalib/utils/animation/Easing.kt

package dev.pandasystems.pandalib.utils.animation

import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

fun interface Easing {
	fun ease(t: Double): Double

	companion object {
		@JvmStatic
		val easeInSine: Easing = Easing { t: Double -> 1 - cos((t * Math.PI) / 2) }
		@JvmStatic
		val easeOutSine: Easing = Easing { t: Double -> sin((t * Math.PI) / 2) }
		@JvmStatic
		val easeInOutSine: Easing = Easing { t: Double -> -(cos(Math.PI * t) - 1) / 2 }

		@JvmStatic
		val easeInQuad: Easing = Easing { t: Double -> t * t }
		@JvmStatic
		val easeOutQuad: Easing = Easing { t: Double -> 1 - (1 - t) * (1 - t) }
		@JvmStatic
		val easeInOutQuad: Easing = Easing { t: Double -> if (t < 0.5) 2 * t * t else 1 - (-2 * t + 2) * (-2 * t + 2) / 2 }

		@JvmStatic
		val easeInCubic: Easing = Easing { t: Double -> t * t * t }
		@JvmStatic
		val easeOutCubic: Easing = Easing { t: Double -> 1 - (1 - t).pow(3.0) }
		@JvmStatic
		val easeInOutCubic: Easing = Easing { t: Double -> if (t < 0.5) 4 * t * t * t else 1 - (-2 * t + 2).pow(3.0) / 2 }

		@JvmStatic
		val easeInQuart: Easing = Easing { t: Double -> t * t * t * t }
		@JvmStatic
		val easeOutQuart: Easing = Easing { t: Double -> 1 - (1 - t).pow(4.0) }
		@JvmStatic
		val easeInOutQuart: Easing = Easing { t: Double -> if (t < 0.5) 8 * t * t * t * t else 1 - (-2 * t + 2).pow(4.0) / 2 }

		@JvmStatic
		val easeInQuint: Easing = Easing { t: Double -> t * t * t * t * t }
		@JvmStatic
		val easeOutQuint: Easing = Easing { t: Double -> 1 - (1 - t).pow(5.0) }
		@JvmStatic
		val easeInOutQuint: Easing = Easing { t: Double -> if (t < 0.5) 16 * t * t * t * t * t else 1 - (-2 * t + 2).pow(5.0) / 2 }

		@JvmStatic
		val easeInExpo: Easing = Easing { t: Double -> if (t == 0.0) 0.0 else 2.0.pow(10 * (t - 1)) }
		@JvmStatic
		val easeOutExpo: Easing = Easing { t: Double -> if (t == 1.0) 1.0 else 1 - 2.0.pow(-10 * t) }
		@JvmStatic
		val easeInOutExpo: Easing =
			Easing { t: Double -> if (t == 0.0) 0.0 else if (t == 1.0) 1.0 else if (t < 0.5) 2.0.pow(20 * t - 10) / 2 else (2 - 2.0.pow(-20 * t + 10)) / 2 }

		@JvmStatic
		val easeInCirc: Easing = Easing { t: Double -> 1 - sqrt(1 - t * t) }
		@JvmStatic
		val easeOutCirc: Easing = Easing { t: Double -> sqrt(1 - (t - 1).pow(2.0)) }
		@JvmStatic
		val easeInOutCirc: Easing = Easing { t: Double -> if (t < 0.5) (1 - sqrt(1 - 4 * t * t)) / 2 else (sqrt(1 - -4 * t * t) + 1) / 2 }

		@JvmStatic
		val easeInBack: Easing = Easing { t: Double ->
			val c1 = 1.70158
			val c3 = c1 + 1
			c3 * t * t * t - c1 * t * t
		}
		@JvmStatic
		val easeOutBack: Easing = Easing { t: Double ->
			val c1 = 1.70158
			val c3 = c1 + 1
			1 + c3 * (t - 1) * (t - 1) * (t - 1) - c1 * (t - 1) * (t - 1)
		}
		@JvmStatic
		val easeInOutBack: Easing = Easing { t: Double ->
			val c1 = 1.70158
			val c2 = c1 * 1.525
			if (t < 0.5)
				((2 * t).pow(2.0) * ((c2 + 1) * 2 * t - c2)) / 2
			else
				((2 * t - 2).pow(2.0) * ((c2 + 1) * (t * 2 - 2) + c2) + 2) / 2
		}

		@JvmStatic
		val easeInElastic: Easing = Easing { t: Double ->
			val c4 = (2 * Math.PI) / 3
			when (t) {
				0.0 -> 0.0
				1.0 -> 1.0
				else -> (-2.0).pow(10 * t - 10) * sin((t * 10 - 10.75) * c4)
			}
		}
		@JvmStatic
		val easeOutElastic: Easing = Easing { t: Double ->
			val c4 = (2 * Math.PI) / 3
			when (t) {
				0.0 -> 0.0
				1.0 -> 1.0
				else -> 2.0.pow(-10 * t) * sin((t * 10 - 0.75) * c4) + 1
			}
		}
		@JvmStatic
		val easeInOutElastic: Easing = Easing { t: Double ->
			val c5 = (2 * Math.PI) / 4.5
			when (t) {
				0.0 -> 0.0
				1.0 -> 1.0
				else -> {
					if (t < 0.5)
						-(2.0.pow(20 * t - 10) * sin((20 * t - 11.125) * c5)) / 2
					else
						(2.0.pow(-20 * t + 10) * sin((20 * t - 11.125) * c5)) / 2 + 1
				}
			}
		}

		@JvmStatic
		val easeOutBounce: Easing = Easing { t: Double ->
			var t = t
			val n1 = 7.5625
			val d1 = 2.75
			return@Easing if (t < 1 / d1) {
				n1 * t * t
			} else if (t < 2 / d1) {
				n1 * (1.5 / d1.let { t -= it; t }) * t + 0.75
			} else if (t < 2.5 / d1) {
				n1 * (2.25 / d1.let { t -= it; t }) * t + 0.9375
			} else {
				n1 * (2.625 / d1.let { t -= it; t }) * t + 0.984375
			}
		}
		@JvmStatic
		val easeInBounce: Easing = Easing { t: Double -> 1 - easeOutBounce.ease(1 - t) }
		@JvmStatic
		val easeInOutBounce: Easing = Easing { t: Double -> if (t < 0.5) (1 - easeOutBounce.ease(1 - 2 * t)) / 2 else (1 + easeOutBounce.ease(2 * t - 1)) / 2 }
	}
}
